src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.hpp

```diff
@@ -0,0 +1,210 @@
+    // This gives maximum capacity as long.MaxValue.
+    // Index equal to long.MaxValue doesn't exist and is reserved for 'NoneMarker' - list terminator.
+    constexpr int MaximumQubitCapacity = INT_MAX;
+
+    // QubitListInSharedArray implements a singly-linked list with "pointers" to the first and the last element stored.
+    // Pointers are the indexes in a shared array. Shared array isn't sotored here because it can be reallocated.
+    // This class maintains status of elements in the list by virtue of linking them as part of this list.
+    // This class doesn't update status of elementes excluded from the list.
+    // This class is small, contains no pointers and relies on default shallow copying/destruction.
+    struct QubitListInSharedArray final
+    {
+    private:
+        int firstElement = NoneMarker;
+        int lastElement = NoneMarker;
+
+    public:
+        // Initialize empty list
+        QubitListInSharedArray() = default;
+
+        // Initialize as a list with sequential elements from minElement to maxElement inclusve.
+        QubitListInSharedArray(int minElement, int maxElement, int* sharedQubitStatusArray);
+
+        bool IsEmpty();
+        void AddQubit(int id, bool addToFront, int* sharedQubitStatusArray);
+        int TakeQubitFromFront(int* sharedQubitStatusArray);
+        void MoveAllQubitsFrom(QubitListInSharedArray& source, int* sharedQubitStatusArray);
+    };
+
+
+    // Restricted reuse area consists of multiple segments. Qubits released in one segment cannot be reused in another.
+    // One restricted reuse area can be nested in a segment of another restricted reuse area.
+    // This class tracks current segment of an area. Previous segments are tracked collectively (not individually).
+    // This class is small, contains no pointers and relies on default shallow copying/destruction.
+    struct RestrictedReuseArea final
+    {
+    public:
+        QubitListInSharedArray FreeQubitsReuseProhibited;
+        QubitListInSharedArray FreeQubitsReuseAllowed;
+
+        RestrictedReuseArea() = default;
+        RestrictedReuseArea(QubitListInSharedArray freeQubits);
+    };
+
+
+
+    // This is NOT a stack! We modify it only by push/pop, but we also iterate over elements.
+    // TODO: Better name?
+    class CRestrictedReuseAreaStack final : public std::vector<RestrictedReuseArea>
+    {
+    public:
+        // No complex scenarios for now. Don't need to support copying/moving.
+        CRestrictedReuseAreaStack() = default;
+        CRestrictedReuseAreaStack(const CRestrictedReuseAreaStack&) = delete;
+        CRestrictedReuseAreaStack& operator = (const CRestrictedReuseAreaStack&) = delete;
+        ~CRestrictedReuseAreaStack() = default;
+
+        void PushToBack(RestrictedReuseArea area);
+        RestrictedReuseArea PopFromBack();
+        RestrictedReuseArea& PeekBack();
+        // TODO: Remove and use size directly?
+        int Count();
+    };
+
```

Comment:
> `class CQubitManagerRestrictedReuse`

FYI: Class names are typically _nouns_. So I intuitively I feel that your code may benefit if you rename the class (and files)
CQubitManagerRestrictedReuse to
C**RestrictedReuse**QubitManager

Reply:
Yes. Actually we don't need any clarifications. It's the only qubit manager we provide. So it's going to be just CQubitManager.